File: solutions/day02.py
from enum import Enum


class Outcome(Enum):
    WIN = 6
    LOSE = 0
    DRAW = 3


class Move(Enum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3

    def outcome(self, other: "Move") -> Outcome:
        return [Outcome.DRAW, Outcome.WIN, Outcome.LOSE][(self.value - other.value) % 3]


class MoveOpp(Enum):
    A = Move.ROCK
    B = Move.PAPER
    C = Move.SCISSORS


class MoveSelf(Enum):
    X = Move.ROCK
    Y = Move.PAPER
    Z = Move.SCISSORS


class NeededOutcome(Enum):
    X = Outcome.LOSE
    Y = Outcome.DRAW
    Z = Outcome.WIN

    def needed_move(self, other: "Move") -> "Move":
        mod = OUTCOME_TO_MOD[self.value]
        return Move((other.value + mod) % 3 or 3)


OUTCOME_TO_MOD: dict[Outcome, int] = {Outcome.WIN: 1, Outcome.DRAW: 0, Outcome.LOSE: 2}


def alpha(inputs: list[str]) -> tuple[int, int]:
    part1 = sum(round_score_part1(*line.split(" ")) for line in inputs)
    part2 = sum(round_score_part2(*line.split(" ")) for line in inputs)
    return part1, part2


def round_score_part1(opponent: str, self_: str) -> int:
    return (
        MoveSelf[self_].value.outcome(MoveOpp[opponent].value).value
        + MoveSelf[self_].value.value
    )


def round_score_part2(opponent: str, outcome: str) -> int:
    needed_outcome = NeededOutcome[outcome]
    return (
        needed_outcome.needed_move(MoveOpp[opponent].value).value
        + needed_outcome.value.value
    )

File: go/day02/day02.go
package day02

import "strings"

func Solve(inputLines []string) (int, int) {
	return partOne(inputLines), partTwo(inputLines)
}

func partOne(inputLines []string) int {
	score := 0

	for _, line := range inputLines {
		moves := strings.Split(line, " ")

		var opponent Move
		switch moves[0] {
		case "A":
			opponent = Rock
		case "B":
			opponent = Paper
		default:
			opponent = Scissors
		}

		var player Move
		switch moves[1] {
		case "X":
			player = Rock
		case "Y":
			player = Paper
		default:
			player = Scissors
		}

		outcome := determineOutcome(player, opponent)
		score += int(outcome) + int(player)
	}

	return score
}

func partTwo(inputLines []string) int {
	score := 0

	for _, line := range inputLines {
		moves := strings.Split(line, " ")

		var opponent Move
		switch moves[0] {
		case "A":
			opponent = Rock
		case "B":
			opponent = Paper
		default:
			opponent = Scissors
		}

		var outcome Outcome
		switch moves[1] {
		case "X":
			outcome = Lose
		case "Y":
			outcome = Draw
		default:
			outcome = Win
		}

		player := determineMove(opponent, outcome)
		score += int(outcome) + int(player)
	}

	return score
}

type Move int

const (
	Rock     Move = 1
	Paper         = 2
	Scissors      = 3
)

type Outcome int

const (
	Win  Outcome = 6
	Lose         = 0
	Draw         = 3
)

func determineOutcome(player Move, opponent Move) Outcome {
	if player == opponent {
		return Draw
	}
	switch player {
	case Rock:
		if opponent == Scissors {
			return Win
		} else {
			return Lose
		}
	case Paper:
		if opponent == Rock {
			return Win
		} else {
			return Lose
		}
	default: // Scissors
		if opponent == Paper {
			return Win
		} else {
			return Lose
		}
	}
}

func determineMove(opponent Move, outcome Outcome) Move {
	switch outcome {
	case Win:
		switch opponent {
		case Rock:
			return Paper
		case Paper:
			return Scissors
		default: // Scissors
			return Rock
		}
	case Lose:
		switch opponent {
		case Rock:
			return Scissors
		case Paper:
			return Rock
		default: // Scissors
			return Paper
		}
	default:
		return opponent
	}
}
